src/Pos/Generator/BlockEvent.hs

```diff
@@ -41,13 +91,334 @@ data BlockRollbackResult
+                        -- Attempt to apply an invalid sequence of blocks.
+                        let tip = NE.last (getOldestFirst blockchain)
+                        blockIndices <- genBlockIndices (tip + 1)
+                        let
+                            blockIndices' =
+                                -- Those block indices are broken by construction
+                                -- because we append the current tip to the end.
+                                -- Sometimes we can violate the maximum bound on
+                                -- block count, but this is fine. See NOTE
+                                -- on reordering corner cases.
+                                over _OldestFirst (Smg.<> pure tip) blockIndices
+                        return . BlkEvApply $ BlockEventApply
+                            { _beaInput    = blockIndices'
+                            , _beaOutValid = BlockApplyFailure
+                            }
+                    AftApplyNonCont -> do
+                        -- Attempt to apply blocks which are not a valid
+                        -- contuniation of our chain.
+                        let
+                            tip  = NE.last (getOldestFirst blockchain)
+                            -- The amount of blocks to skip. One is enough.
+                            skip = 1
+                        blockIndices <- genBlockIndices (tip + 1 + skip)
+                        return . BlkEvApply $ BlockEventApply
+                            { _beaInput    = blockIndices
+                            , _beaOutValid = BlockApplyFailure
+                            }
+            (False, True) -> do
+                -- Success with rollback.
+                len <- getRandomR (1, length blockchain)
+                let
+                    -- 'NE.fromList' is valid here because 'len >= 1'.
+                    select = NE.fromList . NE.take len
+                    blockIndices = toNewestFirst blockchain &
+                        over _NewestFirst select
+                return . BlkEvRollback $ BlockEventRollback
+                    { _berInput    = blockIndices
+                    , _berOutValid = BlockRollbackSuccess
+                    }
+            (False, False) -> do
+                -- Success without rollback (with apply).
+                let tip = NE.last (getOldestFirst blockchain)
+                blockIndices <- genBlockIndices (tip + 1)
+                return . BlkEvApply $ BlockEventApply
+                    { _beaInput = blockIndices
+                    , _beaOutValid = BlockApplySuccess
+                    }
+  where
+    genBlockIndices blockIndexStart = do
+        let blockIndexEndMax = blockIndexStart +
+                               fromIntegral (begp ^. begpBlockCountMax)
+        blockIndexEnd <- getRandomR (blockIndexStart,blockIndexEndMax)
+        return $ OldestFirst . NE.fromList $ [blockIndexStart..blockIndexEnd]
```

Comment:
For example, few lines above you use `genBlockIndices` and put the result into `blockIndices`. And `genBlockIndices` can generate `begpBlockCountMax + 1` indices. But we want at most `begpBlockCountMax` blocks.